torch/_guards.py

```diff
@@ -369,3 +355,18 @@ def tracing(context: TracingContext):
         yield _CURRENT_TRACING_CONTEXT
     finally:
         _CURRENT_TRACING_CONTEXT = old_context
+
+
+"""
+A class representing a pair of duplicate inputs.
+input_pos_a and input_pos_b are input positions we have deduped.
+"""
+
+
+@dataclasses.dataclass
+class DuplicateInputs(GuardEnvExpr):
+    input_source_a: Source
+    input_source_b: Source
+
+    def __post_init__(self):
+        assert self.input_source_a != self.input_source_b
```

Comment:
is this a useful/correct assert?

Imagine the case of  *args, the tuple would have a LocalInputSource and then 2 tensors could have GetAttrSources.

There are all kinds of cases here-
  - 2 same tensors have different GetAttrSources on the same (or different) LocalInputSource
  - 2 different tensors come from the 'same' LocalInputSource (*args)

Reply:
the GetAttrSources should point to different attrs, or different positions.

> 2 same tensors have different GetAttrSources on the same (or different) LocalInputSource

Then they are different sources

> 2 different tensors come from the 'same' LocalInputSource (*args)

For the *args case, we would need to unpack args into per-tensor sources, right?

Reply:
To add more color - we *must* have this assert, because we do not want to get two identical sources here, as that would be a useful guard tantamount to checking $thing is $thing.

Reply:
I guess in the *args case, input_source_a would point to the GetAttrSource, right?  in that case, then the assert will still pass.  Ok, i think this is correct.